### ISSUE TITLE
CSI Driver image v1.2.3 contains critical CVE's (vulnerabilities)

### ISSUE THREAD

**Issue by GiuseppeChiesa-TomTom** [kubernetes-sigs/secrets-store-csi-driver#1026]
**What steps did you take and what happened:**

As part of our gate for accepting container images we have Trivy scan. This scan is reporting CRITICAL vulnerabilities on v1.2.3 images

```
❯ trivy image --ignore-unfixed -s "CRITICAL" k8s.gcr.io/csi-secrets-store/driver:v1.2.3  
2022-08-16T13:09:22.769+0200	INFO	Need to update DB
2022-08-16T13:09:22.769+0200	INFO	Downloading DB...
29.12 MiB / 29.12 MiB [------------------------------------------------------------------------------------------------------------------------------] 100.00% 43.50 MiB p/s 0s
2022-08-16T13:09:25.845+0200	INFO	Detected OS: debian
2022-08-16T13:09:25.845+0200	INFO	Detecting Debian vulnerabilities...
2022-08-16T13:09:25.856+0200	INFO	Number of language-specific files: 0

k8s.gcr.io/csi-secrets-store/driver:v1.2.3 (debian 11.1)
========================================================
Total: 9 (CRITICAL: 9)

+-----------+------------------+----------+-------------------+------------------+---------------------------------------+
|  LIBRARY  | VULNERABILITY ID | SEVERITY | INSTALLED VERSION |  FIXED VERSION   |                 TITLE                 |
+-----------+------------------+----------+-------------------+------------------+---------------------------------------+
| dpkg      | CVE-2022-1664    | CRITICAL | 1.20.9            | 1.20.10          | Vhcalc 0.2.5 updates Dockerfile       |
|           |                  |          |                   |                  | to "python:3.9-slim-buster"           |
|           |                  |          |                   |                  | to include security fixes.            |
|           |                  |          |                   |                  | -->avd.aquasec.com/nvd/cve-2022-1664  |
+-----------+------------------+          +-------------------+------------------+---------------------------------------+
| libc-bin  | CVE-2021-33574   |          | 2.31-13+deb11u2   | 2.31-13+deb11u3  | glibc: mq_notify does                 |
|           |                  |          |                   |                  | not handle separately                 |
|           |                  |          |                   |                  | allocated thread attributes           |
|           |                  |          |                   |                  | -->avd.aquasec.com/nvd/cve-2021-33574 |
+           +------------------+          +                   +                  +---------------------------------------+
|           | CVE-2022-23218   |          |                   |                  | glibc: Stack-based buffer overflow    |
|           |                  |          |                   |                  | in svcunix_create via long pathnames  |
|           |                  |          |                   |                  | -->avd.aquasec.com/nvd/cve-2022-23218 |
+           +------------------+          +                   +                  +---------------------------------------+
|           | CVE-2022-23219   |          |                   |                  | glibc: Stack-based buffer             |
|           |                  |          |                   |                  | overflow in sunrpc clnt_create        |
|           |                  |          |                   |                  | via a long pathname                   |
|           |                  |          |                   |                  | -->avd.aquasec.com/nvd/cve-2022-23219 |
+-----------+------------------+          +                   +                  +---------------------------------------+
| libc6     | CVE-2021-33574   |          |                   |                  | glibc: mq_notify does                 |
|           |                  |          |                   |                  | not handle separately                 |
|           |                  |          |                   |                  | allocated thread attributes           |
|           |                  |          |                   |                  | -->avd.aquasec.com/nvd/cve-2021-33574 |
+           +------------------+          +                   +                  +---------------------------------------+
|           | CVE-2022-23218   |          |                   |                  | glibc: Stack-based buffer overflow    |
|           |                  |          |                   |                  | in svcunix_create via long pathnames  |
|           |                  |          |                   |                  | -->avd.aquasec.com/nvd/cve-2022-23218 |
+           +------------------+          +                   +                  +---------------------------------------+
|           | CVE-2022-23219   |          |                   |                  | glibc: Stack-based buffer             |
|           |                  |          |                   |                  | overflow in sunrpc clnt_create        |
|           |                  |          |                   |                  | via a long pathname                   |
|           |                  |          |                   |                  | -->avd.aquasec.com/nvd/cve-2022-23219 |
+-----------+------------------+          +-------------------+------------------+---------------------------------------+
| libssl1.1 | CVE-2022-1292    |          | 1.1.1k-1+deb11u1  | 1.1.1n-0+deb11u2 | openssl: c_rehash script              |
|           |                  |          |                   |                  | allows command injection              |
|           |                  |          |                   |                  | -->avd.aquasec.com/nvd/cve-2022-1292  |
+           +------------------+          +                   +------------------+---------------------------------------+
|           | CVE-2022-2068    |          |                   | 1.1.1n-0+deb11u3 | openssl: the c_rehash script          |
|           |                  |          |                   |                  | allows command injection              |
|           |                  |          |                   |                  | -->avd.aquasec.com/nvd/cve-2022-2068  |
+-----------+------------------+----------+-------------------+------------------+---------------------------------------+
```

**What did you expect to happen:**

Security scan report no CRITICAL vulnerabilities.

**Environment:**

- Secrets Store CSI Driver version: (use the image tag): v1.2.3
- Kubernetes version: (use `kubectl version`): n/a

**Additional Notes:**

This issue is a follow-up on https://github.com/kubernetes-sigs/secrets-store-csi-driver/issues/1010

I see in the resolution comment you posted the [periodic scan](https://testgrid.k8s.io/sig-auth-secrets-store-csi-driver-periodic#secrets-store-csi-driver-image-scan) you run against your staging environment and from the logs I see indeed the vulnerabilities above are not present.

Nevertheless when we run the check against the official image repository with the tag `k8s.gcr.io/csi-secrets-store/driver:v1.2.3` we get the vulnerabilities.

It's not clear to me how the release process works on your side, but could you please validate the official image is free of the vulnerability reported above?

**Comment by aramase** [comment]
> I see in the resolution comment you posted the [periodic scan](https://testgrid.k8s.io/sig-auth-secrets-store-csi-driver-periodic#secrets-store-csi-driver-image-scan) you run against your staging environment and from the logs I see indeed the vulnerabilities above are not present.

We build the image as part of every scan and some of the packages showed in the CVE get the latest updates as part of the image build. If these CVEs show up in scan job, we will patch it in the main branch and they'll be part of next release.

> Nevertheless when we run the check against the official image repository with the tag k8s.gcr.io/csi-secrets-store/driver:v1.2.3 we get the vulnerabilities.

We don't run scans against released versions because we don't actively patch a released version. CVE patches will be done in main branch and will only be part of next patch/minor release. 

> It's not clear to me how the release process works on your side, but could you please validate the official image is free of the vulnerability reported above?

There were no CVEs at the time of release. All the CVEs at the time of v1.2.3 release were patched
https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/a1839f183ae83fe573a7584817983e1e8e47b9d4/docker/Dockerfile#L31-L34
The image scan job is run at the time of release to ensure that. These CVEs that come up after the release will be patched in the next release. We cut a patch release 1st week of every month to fix CVEs.

Hope this clarifies your questions. If having any of this documented in our release doc would make it clearer, let us know and we can get that added.

**Comment by GiuseppeChiesa-TomTom** [comment]
Thanks for the clear explanation. 
But this does not help me understand how come the same vulnerabilities as `k8s.gcr.io/csi-secrets-store/driver:v1.2.2 ` are present also in `k8s.gcr.io/csi-secrets-store/driver:v1.2.3`, if 1.2.3 included the changes in your main branch.

**Comment by aramase** [comment]
> But this does not help me understand how come the same vulnerabilities as k8s.gcr.io/csi-secrets-store/driver:v1.2.2 are present also in k8s.gcr.io/csi-secrets-store/driver:v1.2.3, if 1.2.3 included the changes in your main branch.

That depends on what is the installed version of the package in the image. For instance, `bsdutils` is `2.36.1-8` in both `v1.2.2` and `v1.2.3`. Only now there is a fixed version for the CVE which is `2.36.1-8+deb11u1`. The same vulnerability will exist in different images based on what version was installed at the time of release. The vulnerabilities in the newer release will be a subset of vulnerabilities of older release but it won't be the exact same.

**Comment by brettjacobson** [comment]
We have this problem with [CVE-2012-2131](https://security-tracker.debian.org/tracker/CVE-2012-2131) which was supposedly resolved in bullseye 1.1.1, but the current [Dockerfile](https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/main/docker/BASEIMAGE) shows the images are built with  bullseye 1.1.0

Will this base image be updated and resolved soon?

**Comment by aramase** [comment]
> We have this problem with [CVE-2012-2131](https://security-tracker.debian.org/tracker/CVE-2012-2131) which was supposedly resolved in bullseye 1.1.1, but the current [Dockerfile](https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/main/docker/BASEIMAGE) shows the images are built with bullseye 1.1.0
> 
> Will this base image be updated and resolved soon?

That file is just a reference for the script, but the actual build uses `bullseye-v1.4.1`: https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/main/docker/Dockerfile#L15. As mentioned in this thread, CVE patches are done every month and will be part of minor or patch release. Older releases will not receive CVE patches.

/close
